src/services/navigate.ts

```diff
@@ -0,0 +1,122 @@
+import { ServiceScene } from '../Components/ServiceScene/ServiceScene';
+import { IndexScene } from '../Components/IndexScene/IndexScene';
+import { ALL_VARIABLE_VALUE } from './variables';
+import { getMetadataService } from './metadata';
+import { locationService } from '@grafana/runtime';
+import {
+  buildServicesUrl,
+  DRILLDOWN_URL_KEYS,
+  PageSlugs,
+  prefixRoute,
+  replaceSlash,
+  ROUTES,
+  ValueSlugs,
+} from './routing';
+import { sceneGraph } from '@grafana/scenes';
+import { UrlQueryMap, urlUtil } from '@grafana/data';
+
+function buildValueBreakdownUrl(label: string, newPath: ValueSlugs, serviceString: string) {
+  if (label === ALL_VARIABLE_VALUE && newPath === ValueSlugs.label) {
+    return prefixRoute(`${PageSlugs.explore}/service/${replaceSlash(serviceString)}/${PageSlugs.labels}`);
+  } else if (label === ALL_VARIABLE_VALUE && newPath === ValueSlugs.field) {
+    return prefixRoute(`${PageSlugs.explore}/service/${replaceSlash(serviceString)}/${PageSlugs.fields}`);
+  } else {
+    return prefixRoute(`${PageSlugs.explore}/service/${replaceSlash(serviceString)}/${newPath}/${replaceSlash(label)}`);
+  }
+}
+
+export function buildBreakdownUrl(path: PageSlugs | string, extraQueryParams?: UrlQueryMap): string {
+  return urlUtil.renderUrl(path, buildBreakdownRoute(extraQueryParams));
+}
+
+export function buildBreakdownRoute(extraQueryParams?: UrlQueryMap): UrlQueryMap {
+  return {
+    ...Object.entries(urlUtil.getUrlSearchParams()).reduce<UrlQueryMap>((acc, [key, value]) => {
+      if (DRILLDOWN_URL_KEYS.includes(key)) {
+        acc[key] = value;
+      }
+
+      return acc;
+    }, {}),
+    ...extraQueryParams,
+  };
+}
+
+/**
+ * Navigate to value breakdown url
+ * @param newPath
+ * @param label
+ * @param serviceScene
+ */
+export function navigateToValueBreakdown(newPath: ValueSlugs, label: string, serviceScene: ServiceScene) {
+  const indexScene = sceneGraph.getAncestor(serviceScene, IndexScene);
+
+  if (indexScene) {
+    const serviceString = indexScene.state.routeMatch?.params.service;
+    if (serviceString) {
+      let urlPath = buildValueBreakdownUrl(label, newPath, serviceString);
+      const fullUrl = buildBreakdownUrl(urlPath);
+
+      // If we're going to navigate, we need to share the state between this instantiation of the service scene
+      if (serviceScene) {
+        const metadataService = getMetadataService();
+        metadataService.setServiceSceneState(serviceScene.state);
+      }
+
+      locationService.push(fullUrl);
+      return;
+    }
+  }
+}
+
+/**
+ * Navigates to the drilldown view specified by the path slug
+ * Note: If the serviceScene is not provided we assume it is not a parent of the calling class, i.e. we're navigating from the service selection view, instead of a drilldown view
+ * Drilldown views should ALWAYS provide the serviceScene state
```

Comment:
The only way to ensure this would be to come to this function and read this comment. Maybe we should split this function into two? One for navigating from the parent scene and another to be used from the service selection page?

Looking at the codebase, the only call that I found was `navigateToBreakdown(ROUTES.logs(service));` and as far as I can see it's not actually "navigating to a breakdown", it's navigating to the logs view.

That would also simplify this function a bit.

Similar to `navigateToIndex()`, maybe `navigateToService()`?

Reply:
Sounds good. In general, for "Drilldown views should ALWAYS provide the serviceScene state" I would have a function where the scene is not optional, and another one when the scene should not be passed, so we don't have to branch execution inside the function and the function-callers knows what to use and what to pass.

Reply:
Yeah good call, I was still waking up when I first read this. I created a much more verbose `navigateToInitialPageAfterServiceSelection` function for this special case where you're navigating from the top-level to the service scene for the first time after selecting a service. Also removed the "breakdown" terminology except in regards to the label value views, and use drilldown when referring to generic sub-pages of the top page.

Reply:
Perfect! Thanks.